Program Security - Updated security introduction lesson (#395)
solana-foundation/developer-content: content/courses/program-security/security-intro.md

---

> title: How to Approach the Program Security Module

Reviewer: ```suggestion
title: How to Approach the Program Security Course
```

objectives:

>   - Understand how to approach the Program Security Module

Reviewer: ```suggestion
  - Understand how to approach the Program Security Course
```

description:
  "Learn how to think intelligently about security for your onchain programs,
  whether developing in Anchor or Native Rust."
---

## Overview

This course aims to introduce you to a range of common security exploits unique
to Solana development. We’ve modeled this course heavily on Coral's
[Sealevel Attacks](https://github.com/coral-xyz/sealevel-attacks) repository.

Program security is covered in our
[Anchor](/content/courses/onchain-development.md) and
[Native Rust](/content/courses/native-onchain-development.md) development
courses to ensure that anyone deploying programs to Mainnet has at least a basic

> understanding of security. If that’s you, the fundamental principles you learned
> in those lessons should help you avoid some common Solana exploits on your own.

Reviewer: ```suggestion
understanding of security. Those courses should help you avoid some common Solana exploits on your own.
```

Reviewer: Nothing to do with your PR, we just need to fix this. All 40 lessons were originally in one giant course!

Author: Sure, got it.

Author: This is added as per comment.

> This module is designed to build on those lessons with two main goals:

Reviewer: ```suggestion
This course builds on those courses with two main goals:
```

1. Expand your awareness of the Solana programming model and highlight areas
   where you need to focus to close security loopholes.

> 2. Introduce you to the tools provided by Anchor to help keep your programs
>    secure.

Reviewer: ```suggestion
2. Introduce you to the tools provided by Anchor to help keep your programs
   secure, and show native Rust users how to implement similar techniques on their own.  
```

> If you completed the Basic Security lesson, the first few topics in this module

Reviewer: ```suggestion
While the first few lessons in this course cover similar topics to the [Anchor course](some link here) or [Program Security lesson](/content/courses/native-onchain-development/program-security.md) in the [Native Course](link here), 
```

Reviewer: Add but also fix these links.

Author: Fixed all the links and made small change in a complete sentence.

> should feel familiar, as they cover similar concepts. However, as you progress,
> you’ll encounter new types of attacks. We encourage you to explore all of them.

Reviewer: ```suggestion
but as you progress,
you’ll encounter new types of attacks. We encourage you to explore all of them.
```

Author: This suggestion on of above comment.

One key point to note is that this course contains more lessons than previous
ones. The lessons are less interdependent, allowing you more flexibility to

> explore topics in the order that suits you best.

Reviewer: make this a callout https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#callout

Author: Added as `info` default callout.

> One key point to note is that this course contains more lessons than previous
> ones. The lessons are less interdependent, allowing you more flexibility to
> explore topics in the order that suits you best.

Reviewer: ```suggestion
Unlike the lessons in other courses, which are in order, you are welcome to explore these lessons in whatever order suits you best.
```

> Initially, we intended to create shorter lessons, but even though each security
> vulnerability may seem "simple," there's a lot to discuss. Consequently, each
> lesson may contain less prose and more code snippets, offering readers the
> flexibility to dive as deep as they want. Ultimately, each lesson remains
> comprehensive, ensuring you gain a solid understanding of the security risks
> discussed.

Reviewer: ```suggestion
Even though each security
vulnerability may seem "simple," there's a lot to discuss. These lessons contain less prose and more code, ensuring you gain a solid understanding of the security risks
discussed.
```

Nothing to do with you but this was a bit sales-y.
